### ISSUE TITLE
`bugprone-unused-return-value` warns about stream insertion operator (`operator<<`)

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#85913]
If you enable the check to warn about all unused returns values  by setting `bugprone-unused-return-value.CheckFunctions` `::.*` it will also report the unchecked usage of `operator<<`.

The following are all common code which is regularly unchecked. So I think it might sense to not warn about these.

```cpp
#include <iostream>

void f()
{
std::cout << "str";
}
```

```cpp
#include <sstream>

void f()
{
std::ostringstream oss;
oss << "str";
}
```

```cpp
#include <fstream>

void f()
{
std::ofstream of("file");
of << "str";
}
```

It should still be reported for the extraction operator as it might indicate actual issues:

```cpp
#include <sstream>

void f()
{
std::istringstream iss("str");
double d;
iss >> d;
}
```
In this case the extraction fails and you need to check the result. See https://godbolt.org/z/KzPor3hMK.

That could be a dedicated check though and this check could just ignore both of them.

CC @HerrCai0907

**Comment by HerrCai0907** [comment]
I don't think it is a good idea to ignore `<<` operator since it has two different semantics. One for stream and the other for bit operator. So the check cannot ignore it simply.

**Comment by NagyDonat** [comment]
The checker could easily check the return type of `operator<<` and silence the reports where it's a stream.

**Comment by NagyDonat** [comment]
I'll create a quick patch that fixes this with a heuristic that silences reports where the return value of `operator<<` is a lvalue reference. This will clearly separate the operators that implement a bitshift (because they return an rvalue) and the ones that are stream operators (because they return an lvalue reference to the stream to enable chaining).

I'll also handle `operator>>` in an analogous way.

**Comment by firewave** [comment]
See https://github.com/danmar/cppcheck/pull/6673 for some context why/how I (ab)use this check.

**Comment by seanm** [comment]
>we could theoretically extend the "configuration language" of the checker to let the user specify that they want to match "::.* minus these stream operators".

Personally, I would really love a way to have warnings for *all* ignored return values *except* for some specific list of functions.

From https://github.com/llvm/llvm-project/issues/85905 it seems it could almost be done, except for the lack of 'negative lookahead patterns' in the regex support.
